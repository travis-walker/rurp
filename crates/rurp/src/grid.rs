use crate::bounds::Bounds;
use euclid::Transform2D;
use geo::Polygon;
use geo_rasterize::{LabelBuilder, Rasterizer};
use ndarray::{prelude::*, ArrayViewMut3};
pub struct WorldSpace;
pub struct ScreenSpace;

#[derive(Debug)]
pub struct Grid {
    data: Array3<f64>,
    x: Array2<f64>,
    y: Array2<f64>,
    bounds: Bounds,
    height: usize,
    width: usize,
    world_height: f64,
    world_width: f64,
    world_to_screen_transform: Transform2D<f64, WorldSpace, ScreenSpace>,
    screen_to_world_transform: Transform2D<f64, ScreenSpace, WorldSpace>,
    nodata: f64,
}
impl Grid {
    /// Get a view of grid data.
    #[must_use]
    pub fn data(&self) -> ArrayView3<f64> {
        self.data.view()
    }

    /// Get a mutable view of grid data.
    pub fn data_mut(&mut self) -> ArrayViewMut3<f64> {
        self.data.view_mut()
    }

    /// Get a view of x values.
    #[must_use]
    pub fn x(&self) -> ArrayView2<f64> {
        self.x.view()
    }

    /// Get a view of y values.
    #[must_use]
    pub fn y(&self) -> ArrayView2<f64> {
        self.y.view()
    }

    /// Get the bounds of the grid.
    #[must_use]
    pub fn bounds(&self) -> Bounds {
        self.bounds.clone()
    }

    /// Get the height of the grid.
    #[must_use]
    pub fn height(&self) -> usize {
        self.height
    }

    /// Get the width of the grid.
    #[must_use]
    pub fn width(&self) -> usize {
        self.width
    }

    /// Get the height of the grid in world units.
    #[must_use]
    pub fn world_height(&self) -> f64 {
        self.world_height
    }

    /// Get the width of the grid in world units.
    #[must_use]
    pub fn world_width(&self) -> f64 {
        self.world_width
    }

    /// Get the transformation from world space to screen space.s
    #[must_use]
    pub fn screen_to_world_transform(&self) -> Transform2D<f64, ScreenSpace, WorldSpace> {
        self.screen_to_world_transform
    }

    /// Get the transformation from screen space to world space.
    #[must_use]
    pub fn world_to_screen_transform(&self) -> Transform2D<f64, WorldSpace, ScreenSpace> {
        self.world_to_screen_transform
    }

    /// Get the nodata value.
    #[must_use]
    pub fn nodata(&self) -> f64 {
        self.nodata
    }
}

impl Grid {
    /// Create a new empty Grid instance with given bounds, resolution, and nodata value.
    #[must_use]
    pub fn empty_from_bounds(bounds: &Bounds, resolution: usize, nodata: f64) -> Self {
        let (left, bottom, right, top) = bounds.clone().into();

        let world_height = top - bottom;
        let world_width = right - left;

        let height = (world_height / resolution as f64).ceil() as usize;
        let width = (world_width / resolution as f64).ceil() as usize;

        let data = Array3::from_elem((height, width, 1), nodata);

        let x_vals = Array1::linspace(left, right, width);
        let x = Array2::from_shape_fn((height, width), |(_, j)| x_vals[j]);

        let y_vals = Array1::linspace(bottom, top, height);
        let y = Array2::from_shape_fn((height, width), |(i, _)| y_vals[i]);

        let world_to_screen_transform: Transform2D<f64, WorldSpace, ScreenSpace> =
            Transform2D::translation(-left, -bottom)
                .then_scale(width as f64 / world_width, height as f64 / world_height);
        let screen_to_world_transform = world_to_screen_transform.inverse().unwrap();

        Grid {
            data,
            x,
            y,
            bounds: bounds.clone(),
            height,
            width,
            world_height,
            world_width,
            world_to_screen_transform,
            screen_to_world_transform,
            nodata,
        }
    }
}

impl Grid {
    fn build_default_rasterizer(&self) -> Rasterizer<f64> {
        let geo_pix_transform = self.world_to_screen_transform.to_untyped();
        LabelBuilder::background(self.nodata)
            .width(self.width)
            .height(self.height)
            .geo_to_pix(geo_pix_transform)
            .build()
            .unwrap()
    }

    /// Rasterize polygons onto the grid.
    pub fn rasterize_polygons(&mut self, polygons: &[Polygon<f64>], polygon_labels: &[f64]) {
        let mut rasterizer = self.build_default_rasterizer();

        polygons
            .iter()
            .zip(polygon_labels)
            .for_each(|(polygon, polygon_label)| {
                rasterizer.rasterize(polygon, *polygon_label).unwrap();
            });

        self.data = rasterizer
            .finish()
            .into_shape((self.height, self.width, 1))
            .unwrap();
    }
}
