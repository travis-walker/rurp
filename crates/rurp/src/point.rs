use ndarray::Array1;

#[derive(Debug)]
pub struct Point {
    pub x: f64,
    pub y: f64,
    pub values: Array1<f64>,
}

impl Point {
    pub fn new(x: f64, y: f64, values: Vec<f64>) -> Self {
        Point {
            x,
            y,
            values: Array1::from_vec(values),
        }
    }
}
