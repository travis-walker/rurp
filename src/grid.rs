use ndarray::{Array1, Array2, Array3};

#[derive(Debug)]
pub struct Grid {
    pub nodata: f64,
    pub data: Array3<f64>,
    pub x: Array2<f64>,
    pub y: Array2<f64>,
}

impl Grid {
    pub fn empty_from_bounds(
        nodata: f64,
        left: f64,
        bottom: f64,
        right: f64,
        top: f64,
        resolution: usize,
    ) -> Self {
        let width = ((right - left) / resolution as f64).ceil() as usize;
        let height = ((top - bottom) / resolution as f64).ceil() as usize;

        let data = Array3::from_elem((height, width, 1), nodata);

        let x_vals = Array1::linspace(left, right, width);
        let x = Array2::from_shape_fn((height, width), |(_, j)| x_vals[j]);

        let y_vals = Array1::linspace(bottom, top, height);
        let y = Array2::from_shape_fn((height, width), |(i, _)| y_vals[i]);

        Grid { nodata, data, x, y }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    fn equivalent(left: f64, right: f64) -> bool {
        left.is_nan() && right.is_nan() || left == right
    }

    #[rstest]
    #[case(4000, 710, 1351, f64::NAN)]
    #[case(8000, 355, 676, f64::NEG_INFINITY)]
    #[case(1200, 2367, 4503, f64::MAX)]
    fn test_from_bounds(
        #[case] resolution: usize,
        #[case] expected_height: usize,
        #[case] expected_width: usize,
        #[case] nodata: f64,
    ) {
        let left = -2221060.;
        let bottom = 523589.;
        let right = 3181702.;
        let top = 3363319.;

        let grid = Grid::empty_from_bounds(nodata, left, bottom, right, top, resolution);

        assert_eq!(grid.data.shape(), [expected_height, expected_width, 1]);

        assert!(equivalent(grid.data[[0, 0, 0]], nodata));
        assert_eq!(grid.x[[0, 0]], left);
        assert_eq!(grid.y[[0, 0]], bottom);

        let max_y = expected_height - 1;
        let max_x = expected_width - 1;
        assert!(equivalent(grid.data[[max_y, max_x, 0]], nodata));
        assert_eq!(grid.x[[max_y, max_x]], right);
        assert_eq!(grid.y[[max_y, max_x]], top);

        let mid_y = expected_height / 2;
        let mid_x = expected_width / 2;
        assert!(equivalent(grid.data[[mid_y, mid_x, 0]], nodata));
        assert!((grid.x[[mid_y, mid_x]] - ((right + left) / 2.)).abs() < resolution as f64);
        assert!((grid.y[[mid_y, mid_x]] - ((top + bottom) / 2.)).abs() < resolution as f64);
    }
}
