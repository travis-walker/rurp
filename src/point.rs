use ndarray::Array1;
pub struct Point {
    pub x: f64,
    pub y: f64,
    pub values: Array1<f64>,
}

impl Point {
    pub fn new(x: f64, y: f64, values: Vec<f64>) -> Self {
        Point {
            x,
            y,
            values: Array1::from_shape_vec(values.len(), values).unwrap(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case(0., 0., vec![1., 2., 3.])]
    #[case(90.1234567, -54.7654321, vec![f64::MAX; 9999])]
    fn test_new(#[case] x: f64, #[case] y: f64, #[case] values: Vec<f64>) {
        let point = Point::new(x, y, values.clone());
        assert_eq!(point.x, x);
        assert_eq!(point.y, y);
        assert_eq!(point.values, Array1::from_vec(values));
    }
}
