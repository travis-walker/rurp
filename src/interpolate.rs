use crate::grid::*;
use crate::point::*;
use std::collections::HashMap;

#[derive(thiserror::Error, Debug)]
pub enum InterpolationError {
    #[error("point is outside grid")]
    PointOutsideOfGrid,
}

pub fn add_point_data_to_grid(
    point_data: &[Point],
    grid: &mut Grid,
) -> Result<(), InterpolationError> {
    let x_width = grid.x[[0, grid.x.shape()[1] - 1]] - grid.x[[0, 0]];
    let x_point_width = x_width / grid.x.shape()[1] as f64;

    let y_height = grid.y[[grid.y.shape()[0] - 1, 0]] - grid.y[[0, 0]];
    let y_point_height = y_height / grid.y.shape()[0] as f64;

    let mut points_grid_points_map: HashMap<(usize, usize), Vec<&Point>> = HashMap::new();

    for point in point_data {
        let point_grid_pos_x = ((point.x - grid.x[[0, 0]]) / x_point_width).floor() as isize;
        let point_grid_pos_y = ((point.y - grid.y[[0, 0]]) / y_point_height).floor() as isize;
        if point_grid_pos_y < 0
            || point_grid_pos_y >= grid.data.shape()[0] as isize
            || point_grid_pos_x < 0
            || point_grid_pos_x >= grid.data.shape()[1] as isize
        {
            return Err(InterpolationError::PointOutsideOfGrid);
        }
        points_grid_points_map
            .entry((point_grid_pos_y as usize, point_grid_pos_x as usize))
            .and_modify(|e| e.push(point))
            .or_insert(vec![point]);
    }
    for ((index_y, index_x), points) in points_grid_points_map {
        grid.data[[index_y, index_x, 0]] =
            points.iter().map(|&p| p.values[0]).sum::<f64>() / points.len() as f64;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array;
    use rstest::rstest;

    #[rstest]
    #[case(0., 0., 10., 10., 1, 30)]
    #[case(-100000., 100000., 100000., 200000., 1000, 9999)]
    #[case::conus(-2221060., 523589., 3181702., 3363319., 4000, 99999)]
    fn test_it_add_point_data_to_grid(
        #[case] left: f64,
        #[case] bottom: f64,
        #[case] right: f64,
        #[case] top: f64,
        #[case] resolution: usize,
        #[case] num_stub_stations: usize,
    ) {
        let x_iter =
            Array::linspace(left, right - resolution as f64 / 2., num_stub_stations).into_iter();
        let y_iter =
            Array::linspace(bottom, top - resolution as f64 / 2., num_stub_stations).into_iter();
        let value_iter = Array::linspace(0., 100., num_stub_stations)
            .into_iter()
            .map(|v| vec![v]);
        let stub_point_data = x_iter
            .zip(y_iter)
            .zip(value_iter)
            .map(|((x, y), values)| Point::new(x, y, values))
            .collect::<Vec<Point>>();
        let mut grid = Grid::empty_from_bounds(f64::NAN, left, bottom, right, top, resolution);

        add_point_data_to_grid(&stub_point_data, &mut grid).unwrap();

        insta::assert_debug_snapshot!(grid);
    }

    #[rstest]
    #[case::x_out_of_bounds(Point::new(2., -1., vec![1.]))]
    #[case::y_out_of_bounds(Point::new(11.0, 2., vec![1.]))]
    fn test_it_returns_error_if_points_are_outside_grid(#[case] err_point: Point) {
        let err_points = vec![err_point];
        let mut grid = Grid::empty_from_bounds(f64::NAN, 0., 0., 10., 10., 1);

        let result = add_point_data_to_grid(&err_points, &mut grid);
        assert!(result.is_err())
    }
}
