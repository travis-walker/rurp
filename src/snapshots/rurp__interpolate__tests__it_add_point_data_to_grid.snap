---
source: src/interpolate.rs
expression: grid
---
Grid {
    nodata: NaN,
    data: [[[5.172413793103448],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [17.241379310344826],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [27.58620689655172],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [37.93103448275862],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [48.275862068965516],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [58.62068965517241],
      [NaN],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [68.9655172413793],
      [NaN],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [79.3103448275862],
      [NaN],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [89.6551724137931],
      [NaN]],
    
     [[NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [NaN],
      [98.27586206896552]]], shape=[10, 10, 1], strides=[10, 1, 1], layout=Cc (0x5), const ndim=3,
    x: [[0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0],
     [0.0, 1.1111111111111112, 2.2222222222222223, 3.3333333333333335, 4.444444444444445, 5.555555555555555, 6.666666666666667, 7.777777777777779, 8.88888888888889, 10.0]], shape=[10, 10], strides=[10, 1], layout=Cc (0x5), const ndim=2,
    y: [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
     [1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112, 1.1111111111111112],
     [2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223, 2.2222222222222223],
     [3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335, 3.3333333333333335],
     [4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445, 4.444444444444445],
     [5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555, 5.555555555555555],
     [6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667, 6.666666666666667],
     [7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779, 7.777777777777779],
     [8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889, 8.88888888888889],
     [10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0]], shape=[10, 10], strides=[10, 1], layout=Cc (0x5), const ndim=2,
}
